eventkitd/utils/resumablecompressed/reader.go

```diff
@@ -0,0 +1,63 @@
+package resumablecompressed
+
+import (
+	"compress/zlib"
+	"errors"
+	"io"
+
+	"github.com/jtolio/eventkit/eventkitd/utils/delimited"
+)
+
+type Reader struct {
+	r       *delimited.Reader
+	current io.ReadCloser
+}
+
+func NewReader(base io.Reader) *Reader {
+	r := delimited.NewReader(base)
+	return &Reader{
+		r: r,
+	}
+}
+
+type readerFunc func(p []byte) (n int, err error)
+
+func (f readerFunc) Read(p []byte) (n int, err error) { return f(p) }
```

Comment:
Instead of using callbacks, write a:

```
type measuredReader struct { io.Reader; Count int64 }
func (r *measuredReader) Read(p []byte) {
    n, err := r.Reader.Read(p)
    r.Count += n
    return n, err
}
```

It'll make the code below slightly easier to read, e.g.:

```
	r := &measuredReader{Reader: r.r}
	current, err := zlib.NewReader(r)

	if err != nil {
			if errors.Is(err, io.ErrUnexpectedEOF) && r.Count == 0 {
				err = io.EOF
```

Reply:
callbacks have always been a point of contention between, say, jeff and i, lol. i think the callback style is way clearer lol.

but okay